new resource "azurerm_synapse_private_link_hub"

owenfarrell opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Closes #12493

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```
$ make acctests SERVICE='synapse' TESTARGS='-run=TestAccSynapsePrivateLinkHub_' TESTTIMEOUT='5m' 
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./azurerm/internal/services/synapse -run=TestAccSynapsePrivateLinkHub_ -timeout 5m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccSynapsePrivateLinkHub_create
=== PAUSE TestAccSynapsePrivateLinkHub_create
=== RUN   TestAccSynapsePrivateLinkHub_update
=== PAUSE TestAccSynapsePrivateLinkHub_update
=== CONT  TestAccSynapsePrivateLinkHub_create
=== CONT  TestAccSynapsePrivateLinkHub_update
--- PASS: TestAccSynapsePrivateLinkHub_create (171.43s)
--- PASS: TestAccSynapsePrivateLinkHub_update (249.49s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/synapse     252.111s
```

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub.go:
```diff
@@ -0,0 +1,167 @@
+	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	existing, err := client.Get(ctx, resourceGroup, name)
```
We generally prefer to generate a resource ID first and use it through the Create function

```suggestion
	subscriptionId := meta.(*clients.Client).Account.SubscriptionId
	id := parse.NewPrivateLinkHubID(subscriptionId, d.Get("resource_group_name").(string), d.Get("name").(string))

	existing, err := client.Get(ctx, id.ResourceGroup, id.Name)
```

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub.go:
```diff
@@ -0,0 +1,167 @@
+	existing, err := client.Get(ctx, resourceGroup, name)
+	if err != nil {
+		if !utils.ResponseWasNotFound(existing.Response) {
+			return fmt.Errorf("checking for present of existing Synapse Private Link Hub %q (Resource Group %q): %+v", name, resourceGroup, err)
```
The id struct has a useful String method which can be used for simpler and more consistent error messages

```suggestion
			return fmt.Errorf("checking for presence of %s: %+v", id, err)
```

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub.go:
```diff
@@ -0,0 +1,167 @@
+		if !utils.ResponseWasNotFound(existing.Response) {
+			return fmt.Errorf("checking for present of existing Synapse Private Link Hub %q (Resource Group %q): %+v", name, resourceGroup, err)
+		}
+	}
+	if existing.ID != nil && *existing.ID != "" {
+		return tf.ImportAsExistsError("azurerm_synapse_private_link_hub", *existing.ID)
```
Using the generated id here will help avoid issues due to malformed API responses

```suggestion
	} else {
		return tf.ImportAsExistsError("azurerm_synapse_private_link_hub", id.ID())
```

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub.go:
```diff
@@ -0,0 +1,167 @@
+		Tags:     tags.Expand(d.Get("tags").(map[string]interface{})),
+	}
+
+	_, err = client.CreateOrUpdate(ctx, privateLinkHubInfo, resourceGroup, name)
```
By generating the id first, we can use it everywhere we need its fields

```suggestion
	_, err = client.CreateOrUpdate(ctx, privateLinkHubInfo, id.ResourceGroup, id.Name)
```

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub.go:
```diff
@@ -0,0 +1,167 @@
+
+	_, err = client.CreateOrUpdate(ctx, privateLinkHubInfo, resourceGroup, name)
+	if err != nil {
+		return fmt.Errorf("creating Synapse Private Link Hub %q (Resource Group %q): %+v", name, resourceGroup, err)
```
As above, simpler error construction

```suggestion
		return fmt.Errorf("creating %s: %+v", id, err)
```

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub.go:
```diff
@@ -0,0 +1,167 @@
+		return fmt.Errorf("creating Synapse Private Link Hub %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	subscriptionId := meta.(*clients.Client).Account.SubscriptionId
+	id := parse.NewPrivateLinkHubID(subscriptionId, resourceGroup, name)
```
(moved further up)
```suggestion
```

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub.go:
```diff
@@ -0,0 +1,167 @@
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("retrieving Synapse Private Link Hub %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
```
Also here
```suggestion
		return fmt.Errorf("retrieving %s: %+v", id, err)
```

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub.go:
```diff
@@ -0,0 +1,167 @@
+		return err
+	}
+
+	if d.HasChanges("tags") {
```
Can use d.HasChange if we're only checking one property
```suggestion
	if d.HasChange("tags") {
```

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub.go:
```diff
@@ -0,0 +1,167 @@
+
+		_, err := client.Update(ctx, privateLinkHubPatchInfo, id.ResourceGroup, id.Name)
+		if err != nil {
+			return fmt.Errorf("updating Synapse Private Link Hub %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
```
```suggestion
			return fmt.Errorf("updating %s: %+v", id, err)
```

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub.go:
```diff
@@ -0,0 +1,167 @@
+
+	future, err := client.Delete(ctx, id.ResourceGroup, id.Name)
+	if err != nil {
+		return fmt.Errorf("deleting Synapse Private Link Hub %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
```
```suggestion
		return fmt.Errorf("deleting %s: %+v", id, err)
```

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub.go:
```diff
@@ -0,0 +1,167 @@
+	// sometimes the waitForCompletion rest api will return 404
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		if !response.WasNotFound(future.Response()) {
+			return fmt.Errorf("waiting for Synapse Private Link Hub %q (Resource Group %q) to be deleted: %+v", id.Name, id.ResourceGroup, err)
```
```suggestion
			return fmt.Errorf("waiting for deletion of %s: %+v", id, err)
```

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub_test.go:
```diff
@@ -0,0 +1,116 @@
+
+type SynapsePrivateLinkHubResource struct{}
+
+func TestAccSynapsePrivateLinkHub_create(t *testing.T) {
```
We prefer to align the test name and the config method name where there's a 1:1 correspondence

```suggestion
func TestAccSynapsePrivateLinkHub_basic(t *testing.T) {
```

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub_test.go:
```diff
@@ -0,0 +1,116 @@
+func (r SynapsePrivateLinkHubResource) basic(data acceptance.TestData) string {
+	template := r.template(data)
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_synapse_private_link_hub" "test" {
+  name                = "acctestsw%d"
```
Using indexed verbs will make it easier to read and reduce repetition if config needs to be added at a later date
```suggestion
%[1]s

resource "azurerm_synapse_private_link_hub" "test" {
  name                = "acctestsw%[2]d"
```

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub_test.go:
```diff
@@ -0,0 +1,116 @@
+func (r SynapsePrivateLinkHubResource) withUpdateFields(data acceptance.TestData) string {
+	template := r.template(data)
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_synapse_private_link_hub" "test" {
+  name                = "acctestsw%d"
```
As above
```suggestion
%[1]s

resource "azurerm_synapse_private_link_hub" "test" {
  name                = "acctestsw%[2]d"
```

manicminer commented on website/docs/r/synapse_private_link_hub.html.markdown:
````diff
@@ -0,0 +1,51 @@
+
+```hcl
+resource "azurerm_synapse_private_link_hub" "example" {
+  name     = "example-resources"
+  location = "West Europe"
````
It looks like `resource_group_name` is missing here

```suggestion
  name                = "example-resource"
  resource_group_name = "example-rg"
  location            = "West Europe"
```

manicminer commented on website/docs/r/synapse_private_link_hub.html.markdown:
```diff
@@ -0,0 +1,51 @@
+The following arguments are supported:
+
+* `name` - (Required) The name which should be used for this Synapse Private Link Hub. Changing this forces a new Synapse Private Link Hub to be created.
+
```
`resource_group_name` and `location` should be listed here

```suggestion

* `resource_group_name` - (Required) The name of the resource group in which to create the Synapse Private Link Hub. Changing this forces a new resource to be created.

* `location` - (Required) Specifies the Azure location where the Synapse Private Link Hub exists. Changing this forces a new resource to be created.

```

manicminer commented on website/docs/r/synapse_private_link_hub.html.markdown:
````diff
@@ -0,0 +1,51 @@
+Synapse Private Link Hub can be imported using the `resource id`, e.g.
+
+```shell
+terraform import azurerm_synapse_private_link_hub.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/group1/providers/Microsoft.Synapse/privateLinkHubs/privateLinkHub1
````
We should double check that this format is correct once the id template is added to `resourceids.go`

manicminer reviewed (changes requested):
Hi @owenfarrell, many thanks for this PR, this is off to a great start! I have made some suggestions inline below, most of which relate to some more recent changes/conventions within the provider so may not be immediately obvious when looking at some pre-existing resources for inspiration.

Did you generate the id parser/validator structs and functions? I noticed that the id template for this resource still needs to be added to `resourceids.go` (in the same directory as the resource and acceptance test files). Once it's added there, you can run `make generate` to (re)generate these - although given that the tests are passing there probably won't be any changes.

Your tests are great, I made a few small suggestions and we'll also need to add an import check test - this should create a resource, then try to create another identical one and verify that the import check in the Create function is working as expected.

If you can take a look at these, and all the tests continue to pass, this should be good to merge. Thanks again!

manicminer commented on azurerm/internal/services/synapse/synapse_private_link_hub_test.go:
```diff
@@ -0,0 +1,116 @@
+		if utils.ResponseWasNotFound(resp.Response) {
+			return utils.Bool(false), nil
+		}
+		return nil, fmt.Errorf("retrieving Synapse Private Link Hub %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
```
```suggestion
		return nil, fmt.Errorf("retrieving %s: %+v", id, err)
```

owenfarrell commented:
@manicminer Thanks for the review!

I folded in all of your proposed changes, updated `resources.go` to generate the parsing/validation code (my copypasta turned out to be accurate), and added an `_requiresImport` test.

Everything still checks out okay - let me know if there's anything else.

```
$ make acctests SERVICE='synapse' TESTARGS='-run=TestAccSynapsePrivateLinkHub_' TESTTIMEOUT='5m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./azurerm/internal/services/synapse -run=TestAccSynapsePrivateLinkHub_ -timeout 5m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccSynapsePrivateLinkHub_basic
=== PAUSE TestAccSynapsePrivateLinkHub_basic
=== RUN   TestAccSynapsePrivateLinkHub_requiresImport
=== PAUSE TestAccSynapsePrivateLinkHub_requiresImport
=== RUN   TestAccSynapsePrivateLinkHub_update
=== PAUSE TestAccSynapsePrivateLinkHub_update
=== CONT  TestAccSynapsePrivateLinkHub_basic
=== CONT  TestAccSynapsePrivateLinkHub_update
=== CONT  TestAccSynapsePrivateLinkHub_requiresImport
--- PASS: TestAccSynapsePrivateLinkHub_requiresImport (181.74s)
--- PASS: TestAccSynapsePrivateLinkHub_basic (184.69s)
--- PASS: TestAccSynapsePrivateLinkHub_update (210.74s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/synapse     213.359s
```

manicminer reviewed (approved):
Thanks @owenfarrell, this LGTM 🚀 

<img width="607" alt="Screenshot 2021-07-15 at 14 51 38" src="https://user-images.githubusercontent.com/251987/125799641-b342da68-512f-4a54-b9ba-c69c926cfb95.png">
